### DISCUSSION TITLE
TypeMismatch when trying to save a set of models

### DISCUSSION THREAD

**Discussion by steven-peralta** [dynamoose/dynamoose#1307] [Q&A]
I'm working on migrating an existing project I have from using Mongoose to Dynamoose but I'm having trouble when it comes to creating documents for a model whose schema is a Set of another model.

here's an example:

series model:
```typescript
class Series extends Document {
  id: number;
  name: string;
}

const SeriesSchema = new dynamoose.Schema({
  id: Number,
  name: String
});

const SeriesModel = dynamoose.model<Series>('Series', SeriesSchema);
```

character model:
```typescript
class Character extends Document {
  id!: number;
  name!: string;
  appearances?: Set<Series>;
}

const CharacterSchema = new dynamoose.Schema({
  id: Number,
  name: String,
  appearances: { type: Set, schema: [SeriesModel] }
});

const CharacterModel = dynamoose.model<Character>('Character', CharacterSchema);
```

later in my script I have a function that pulls the Series data from my legacy MongoDB instance and massages the data into DynamoDB like so:
```typescript
// legacy code import
import LegacySeries, { seriesModel } from '$db/models/Series';

const series: Series[] = [];
const legacySeries: LegacySeries[] = await seriesModel.find().lean();
legacySeries.forEach(doc => {
  const { id, name } = doc;
  series[id] = new StudioModel({ id, name });
});
```
this works fine, the issue is when it comes to converting my Character models:
```typescript
// legacy code import
import LegacyCharacter, { characterModel } from '$db/models/Character';

const characters: Character[] = [];
const legacyCharacters: LegacyCharacters[] = await characterModel.find().lean();
legacyCharacters.forEach(doc => {
  const { id, name, appearances: legacyAppearances } = doc;
  const appearances: Set<Series> = new Set();
  // populate our set using references of the previously created series above
  if (legacyAppearances && legacyAppearances.length > 0) {
    legacyAppearances.forEach(seriesId => appearances.add(series[seriesId as number]))
  }
  characters[id] = new CharacterModel({
    id,
    name,
    appearances
  });
});
```

when it comes to saving the Character models, I get the following error:
```
TypeMismatch: Expected appearances to be of type Series Set, instead found type object.
    at checkTypeFunction (/Users/steven/Development/Webstorm/rosiebot/node_modules/dynamoose/lib/Document.ts:332:11)
    at Array.map (<anonymous>)
    at Function.Document.objectFromSchema (/Users/steven/Development/Webstorm/rosiebot/node_modules/dynamoose/lib/Document.ts:353:111)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at Document.toDynamo (/Users/steven/Development/Webstorm/rosiebot/node_modules/dynamoose/lib/Document.ts:513:17)
    at /Users/steven/Development/Webstorm/rosiebot/node_modules/dynamoose/lib/Model/index.ts:590:15
    at async Promise.all (index 0)
    at /Users/steven/Development/Webstorm/rosiebot/node_modules/dynamoose/lib/Model/index.ts:588:18
    at /Users/steven/Development/Webstorm/rosiebot/node_modules/dynamoose/lib/Model/index.ts:611:22
```

I'm not sure why this is happening. How are you supposed to save a set of models with this sort of schema? I thought it would be as straight forward as the code I have above.

**Comment by fishcharlie** [comment]
@k9rosie What happens when you make `appearances` an array? I know that isn't what you want, but try that and see if it works without errors or problems. If it does, it's likely a bug and needs to be reported in the issues section of this repo.

Additionally, `characterModel.find().lean();` isn't valid at all in Dynamoose, so is there custom code you have for handling that? Can you show that code?

Finally, where is your save code that is throwing this error?

**Reply by steven-peralta** [reply]
The code works when making `appearances` an array.

> Additionally, `characterModel.find().lean();` isn't valid at all in Dynamoose, so is there custom code you have for handling that? Can you show that code?

That's an API call to Mongoose that gets the legacy character data from MongoDB. Like I said, I'm converting legacy Mongoose documents to my Dynamoose models.

This is my save code. `chunkedCharacters` is a 2D array of newly instanced Character models divided into chunks of 25 items each for `batchPut`:

```typescript
      console.log('Saving characters...');
      let i = 0;
      const promises = characterChunks.map(async (chk) => {
        await CharacterModel.batchPut(chk);
        i++;
        console.log(`Chunk ${i}/${promises.length} saved`);
      });
      await Promise.all(promises);
```

**Reply by fishcharlie** [reply]
> The code works when making `appearances` an array.

@k9rosie Awesome. Create a MCVE and submit an issue to this repo if you could 😃.

Should be as simple as saying using native JS set doesn't work with saving using Dynamoose Set Type. Be sure to mention that using an array works, and link to this discussion.

**Reply by steven-peralta** [reply]
created #1308
